rts/motoko-rts-tests/src/gc.rs

```diff
@@ -0,0 +1,274 @@
+        write!(
+            &mut error_message,
+            "Reachable objects missing in the post-GC heap: {:?}",
+            missing_objects,
+        )
+        .unwrap();
+    }
+
+    if !extra_objects.is_empty() {
+        if !error_message.is_empty() {
+            error_message.push('\n');
+        }
+
+        write!(
+            &mut error_message,
+            "Unreachable objects seen in the post-GC heap: {:?}",
+            extra_objects,
+        )
+        .unwrap();
+    }
+
+    if !error_message.is_empty() {
+        panic!("{}", error_message);
+    }
+}
+
+fn compute_reachable_objects(
+    roots: &[ObjectIdx],
+    heap: &HashMap<ObjectIdx, Vec<ObjectIdx>>,
+) -> HashSet<ObjectIdx> {
+    let mut closure: HashSet<ObjectIdx> = roots.iter().copied().collect();
+    let mut work_list: Vec<ObjectIdx> = roots.iter().copied().collect();
+
+    while let Some(next) = work_list.pop() {
+        let pointees = heap
+            .get(&next)
+            .unwrap_or_else(|| panic!("Object {} is in the work list, but not in heap", next));
+
+        for pointee in pointees {
+            if closure.insert(*pointee) {
+                work_list.push(*pointee);
+            }
+        }
+    }
+
+    closure
+}
+
+impl GC {
+    fn run(&self, mut heap: MotokoHeap) {
+        let heap_base = heap.heap_base_address() as u32;
+        let static_roots = skew(heap.static_root_array_address());
+        let closure_table_address = heap.closure_table_address() as *mut SkewedPtr;
+
+        let heap_1 = heap.clone();
+        let heap_2 = heap.clone();
+
+        match self {
+            GC::Copying => {
+                unsafe {
+                    copying_gc_internal(
+                        &mut heap,
+                        heap_base,
+                        // get_hp
+                        || heap_1.heap_ptr_address(),
+                        // set_hp
+                        move |hp| heap_2.set_heap_ptr_address(hp as usize),
+                        static_roots,
+                        closure_table_address,
+                        // note_live_size
+                        |_live_size| {},
+                        // note_reclaimed
+                        |_reclaimed| {},
+                    );
+                }
+            }
+
+            GC::MarkCompact => {
+                unsafe {
+                    compacting_gc_internal(
+                        &mut heap,
+                        heap_base,
+                        // get_hp
+                        || heap_1.heap_ptr_address(),
+                        // set_hp
+                        move |hp| heap_2.set_heap_ptr_address(hp as usize),
+                        static_roots,
+                        closure_table_address,
+                        // note_live_size
+                        |_live_size| {},
+                        // note_reclaimed
+                        |_reclaimed| {},
+                    );
+                }
+            }
+        }
```

Comment:
So the problem here is `compacting_gc_internal` and `copying_gc_internal` functions have generic parameters, so they're like C++ templates rather than OCaml/Haskell functions with type parameters. We can't return them from functions for two reasons (1) type system not powerful enough for higher rank types, so we can't return a value of type `forall x y z . ...` (2) the runtime does not support values with type parameters (which would require monomorphising in runtime). I also mentioned this briefly in the `GC` type, copying here:

```rust
/// Enum for the GC implementations. GC functions are generic so we can't put them into arrays or
/// other data types, we use this type instead.
#[derive(Debug, Clone, Copy)]
pub enum GC {
    Copying,
    MarkCompact,
}
```